frontend/catalyst/device/qjit_device.py

```diff
@@ -520,10 +524,37 @@ def _measurement_transform_program(self):
         elif not supports_sum_observables:
             measurement_program.add_transform(split_to_single_terms)
 
-        if self.measurement_processes in [{"Sample"}, {"Counts", "Sample"}]:
-            measurement_program.add_transform(measurements_from_samples, self.wires)
-        if self.measurement_processes == {"Counts"}:
-            measurement_program.add_transform(measurements_from_counts, self.wires)
+        if not self.observables:
+            if not split_non_commuting in measurement_program:
+                # this *should* be redundant, a TOML that doesn't have observables should have
+                # a False non_commuting_observables flag, but we aren't enforcing that
+                measurement_program.add_transform(split_non_commuting)
+            if "Sample" in self.measurement_processes:
+                measurement_program.add_transform(measurements_from_samples, self.wires)
+            elif "Counts" in self.measurement_processes:
+                measurement_program.add_transform(measurements_from_counts, self.wires)
+            else:
+                raise RuntimeError("The device does not support observables or sample/counts")
+        elif not {"PauliX", "PauliY", "PauliZ", "Hadamard"}.issubset(self.observables):
+            if not split_non_commuting in measurement_program:
+                # the device might support non commuting measurements but not all the
+                # Pauli + Hadamard observables, so here it is needed
+                measurement_program.add_transform(split_non_commuting)
+            _obs_dict = {
+                "PauliX": qml.X,
+                "PauliY": qml.Y,
+                "PauliZ": qml.Z,
+                "Hadamard": qml.Hadamard,
+            }
+            # checking which base observables are unsupported and need to be diagonalized
+            supported_observables = {"PauliX", "PauliY", "PauliZ", "Hadamard"}.intersection(
+                self.observables
+            )
+            supported_observables = [_obs_dict[obs] for obs in supported_observables]
+
+            measurement_program.add_transform(
+                diagonalize_measurements, supported_base_obs=supported_observables
+            )
```

Comment:
What if the the device support observables but only measurement and counts? Should we not add that those cases at the end? (maybe I missing something)

```
            if "Sample" in self.measurement_processes:
                measurement_program.add_transform(measurements_from_samples, self.wires)
            elif "Counts" in self.measurement_processes:
                measurement_program.add_transform(measurements_from_counts, self.wires)
            else:
                raise RuntimeError("The device does not support observables or 
```

Reply:
I guess for the time being we could, to allow broader support? My impression from the conversation the other day was that we don't want to cover the case where observables _are_ supported with a transform that converts to the computational basis - and that long term the "counts only observables ok" scenario should be covered by the not-yet-implemented transform that David talked about, that leaves the observables in-tact but changes the measurement type to `counts`. And maybe does the conversion conditionally, so it isn't an all-or-nothing conversion, but rather a "fall back to counts for all unsupported MPs" transform.

But currently catalyst doesn't support counts with observables, so that's not something that will happen in the near-term. I could apply this in that case for now, if we want to make sure that its supported in the meantime.

Reply:
I was wondering if I should combine these two transforms and rename them something like `measure_in_computational_basis`, with a kwarg to indicate counts vs samples, to more clearly distinguish this transform from some potential future `unsupported_measurements_to_counts` transform that would just selectively turn all unsupported MPs to counts without modifying observables. Would that be a clearer thing to do? @dime10

Reply:
I think that sounds great 👍 I guess Romain wasn't there when we discussed some of the intricacies so maybe we should document the current state/approach and what is/isn't supported in some comments? That would help everyone align I think

Reply:
Ok sounds good thanks for the answer!